### PR TITLE
Import all the things!

### LINKED ISSUE #581: right click- open with- prism launcher (zip/mrpack) doesn't work on linux
### Operating System

Linux

### Version of Prism Launcher

develop

### Version of Qt

qt 5.15.7

### Description of bug

right click- open with- prism launcher should open the modpack/rpack installing dialog with zip/mrpack, but it just gives... nothing

### Steps to reproduce

- right click- open with- prism launcher on a zip or a mrpack
- see prism launching normally

### Suspected cause

idk but when opening on a terminal 

prismlauncher path/to/file

it gives

QLocalSocket::setServerName() called while not in unconnected state

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### LINKED ISSUE #700: Windows File Association Bug
### Operating System

Windows

### Version of Prism Launcher

6.1

### Version of Qt

Qt 6.4.0

### Description of bug

Expected : when opening zip / mrpack using prism, it will open the prism import page, and put the correct file path.
Happened : The file path is escaped, and it cannot be imported

### Steps to reproduce

- Download any modpack (rmpack / zip)
- Right click > Open With > Minecraft Modpack
- See the bug

### Suspected cause

For example, the mrpack file is in
`C:\Users\user\Downloads\CreateExtra.mrpack`, the result would be `c:%5CUsers%5Cuser%5CDownloads%5CCreateExtra.mrpack`.
See the problem?
Yes, it's because it's escaping the `\` to be `%5C`.
I think this is a Windows problem.
As you can see in the registry (`HKEY_CLASSES_ROOT\PrismLauncher.App\shell\open\command`)
There is nothing wrong. It just escapes the backslash.

So there is 2 way to solve this problem
1. When prism got the `-l` argument, it searches for any URL Escape (%XX) and will change it to the corresponding character (ex. `%20` ⇒ ` `, `%5C` ⇒ `\`) *Also capitalize the drive letter to make it nicer*
2. or, Prism should handle URL Escaped file path. *But it will look ugly because of the `%` stuff to the user.*

The first one is probably better.

Note: It seems Windows just escape `\`, but does not escape any other special character (ex. $#@!)

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### PR DESCRIPTION
**Pull request by Ryex**
Based off #576

Fixes #581
Fixes #700

  * Adds processing for any number of files passed via cmdline without other flags to be imported.
  * allows the `-I` flag to be passed multiple times
  * files passed with the `-I` flag and though tacked on at the end without flags are treated the same are functionally identical
  * all paths passes are identified with the work from #576.
    * for each file a dialog asks what instance they wish to import the file to
    * if the file is not identified it attempts to import it as an instance
    * only data packs are ignored as we have no real handling of them yet.

### CODE REVIEW

**Comment by Ryex** [comment]
with the latest commit this should fix #700 as well

**Comment by DioEgizio** [comment]
This should be rebased

File: launcher/minecraft/mod/tasks/LocalResourceParse.cpp [hunk 1]
```diff
@@ -57,4 +70,12 @@ PackedResourceType identify(QFileInfo file){
...
 }
+
+QString getPackedTypeName(PackedResourceType type) {
+    return s_packed_type_names.constFind(type).value();
 }
+
+}
+
+
+
```
Comment (flowln) [inline launcher/minecraft/mod/tasks/LocalResourceParse.cpp]: leave only one newline :P

File: launcher/ui/MainWindow.cpp [hunk 2]
```diff
@@ -1826,31 +1827,52 @@ void MainWindow::droppedURLs(QList<QUrl> urls)
...
         QFileInfo localFileInfo(localFileName);
 
-        bool isResourcePack = ResourcePackUtils::validate(localFileInfo);
-        bool isTexturePack = TexturePackUtils::validate(localFileInfo);
+        auto type = ResourceUtils::identify(localFileInfo);
+
+        // bool is_resource = type;
 
-        if (!isResourcePack && !isTexturePack) {  // probably instance/modpack
+        if (!(ResourceUtils::ValidResourceTypes.count(type) > 0)) {  // probably instance/modpack
```
Comment (flowln) [inline launcher/ui/MainWindow.cpp]:
```suggestion
        if (!ResourceUtils::ValidResourceTypes.contains(type)) {  // probably instance/modpack
```

File: launcher/ui/dialogs/ImportResourceDialog.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+namespace Ui {
+class ImportResourceDialog;
+}
+
+class ImportResourceDialog : public QDialog {
+    Q_OBJECT
+
+   public:
+    explicit ImportResourceDialog(QString file_path, PackedResourceType type, QWidget* parent = 0);
+    ~ImportResourceDialog();
```
Comment (flowln) [inline launcher/ui/dialogs/ImportResourceDialog.h]:
```suggestion
    explicit ImportResourceDialog(QString file_path, PackedResourceType type, QWidget* parent = nullptr);
    ~ImportResourceDialog() override;
```

File: launcher/ui/dialogs/ImportResourceDialog.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    InstanceProxyModel* proxyModel;
```
Comment (flowln) [inline launcher/ui/dialogs/ImportResourceDialog.h]: can this go into the private section? :eyes:

**Review by flowln — COMMENTED** [review]
this looks good to me, just did some nitpicks :3

**Comment by flowln** [comment]
btw, 47381a1 has some typos in the description :P

File: launcher/ui/MainWindow.cpp (hunk 2 above)
Comment (Ryex) [reply launcher/ui/MainWindow.cpp]: `std::set.contains` is only in C++20 aren't we limiting ourselves to c++17?

Comment (flowln) [reply launcher/ui/MainWindow.cpp]: oh
i didn't notice that :clown_face: 

_pretend i never said anything_

**Comment by Ryex** [comment]
fixed.
also amended 47381a1 to a4870d4

**Review by flowln — APPROVED** [review]
lgtm, thanks!

File: launcher/Application.cpp [hunk 5]
```diff
@@ -259,9 +259,18 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
...
     m_liveCheck = parser.isSet("alive");
-    m_zipToImport = parser.value("import");
+    
     m_instanceIdToShowWindowOf = parser.value("show");
 
+    for (auto zip_path : parser.values("import")){
+        m_zipsToImport.append(QUrl::fromLocalFile(QFileInfo(zip_path).absoluteFilePath()));
+    }
+
+    for (auto zip_path : parser.positionalArguments()){ // treat unspesified positional arguments as import urls
```
Comment (Scrumplex) [inline launcher/Application.cpp]:
```suggestion
    // treat unspecified positional arguments as import urls
    for (auto zip_path : parser.positionalArguments()) {
```

File: launcher/ui/MainWindow.cpp [hunk 6]
```diff
@@ -1813,10 +1812,12 @@ void MainWindow::on_actionAddInstance_triggered()
     addInstance();
 }
 
-void MainWindow::droppedURLs(QList<QUrl> urls)
+void MainWindow::processURLs(QList<QUrl> urls)
 {
     // NOTE: This loop only processes one dropped file!
     for (auto& url : urls) {
+        qDebug() << "Processing :" << url;
```
Comment (Scrumplex) [inline launcher/ui/MainWindow.cpp]: nit:
```suggestion
        qDebug() << "Processing" << url;
```

File: launcher/ui/MainWindow.cpp [hunk 7]
```diff
@@ -1826,31 +1827,52 @@
...
 
-        auto localFileName = url.toLocalFile();
+        auto localFileName = QDir::toNativeSeparators(url.toLocalFile()) ;
         QFileInfo localFileInfo(localFileName);
 
-        bool isResourcePack = ResourcePackUtils::validate(localFileInfo);
-        bool isTexturePack = TexturePackUtils::validate(localFileInfo);
+        auto type = ResourceUtils::identify(localFileInfo);
+
+        // bool is_resource = type;
```
[the remark this answers is not in the document]
Comment (Scrumplex) [reply launcher/ui/MainWindow.cpp]: indeed

File: launcher/ui/MainWindow.cpp (hunk 2 above)
Comment (Scrumplex) [inline launcher/ui/MainWindow.cpp]:
```suggestion
        if (ResourceUtils::ValidResourceTypes.count(type) == 0) {  // probably instance/modpack
```

**Comment by Ryex** [comment]
fixed

**Review by Scrumplex — APPROVED** [review]
Thanks!

**Pull request merged**